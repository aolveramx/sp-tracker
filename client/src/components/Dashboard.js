import { useEffect, useState } from 'react'
import Player from './Player'
import SpotifyWebApi from 'spotify-web-api-node'
import SearchResults from './SearchResults'
import useAuth from '../hooks/useAuth'

import { Grid2 } from './styles/Dashboard/Grid2.styled'
import { Container } from './styles/layout/Container.styled'
import { Form } from './styles/Dashboard/Form.styled'
import { SearchBar } from './styles/Dashboard/SearchBar.styled'
import { TrackSearchResultContainer } from './styles/Dashboard/TrackSearchResultContainer.styled'

const spotifyApi = new SpotifyWebApi({
  clientId: process.env.REACT_APP_CLIENT_ID,
})

const Dashboard = ({ code }) => {
  const accessToken = useAuth(code)
  const [search, setSearch] = useState('')
  const [searchResults, setSearchResults] = useState([])
  const [playingTrack, setPlayingTrack] = useState()

  const chooseTrack = track => {
    setPlayingTrack(track)
    setSearch('')
  }

  useEffect(() => {
    if (!accessToken) return
    spotifyApi.setAccessToken(accessToken)
  }, [accessToken])

  useEffect(() => {
    if (!search) return setSearchResults([])
    if (!accessToken) return

    let cancel = false

    spotifyApi.searchTracks(search).then(res => {
      if (cancel) return
      setSearchResults(
        res.body.tracks.items.map(track => {
          const smallesAlbumImage = track.album.images.reduce(
            (smallest, image) => {
              if (image.height < smallest.height) return image
              return smallest
            },
            track.album.images[0]
          )

          return {
            artist: track.artists[0].name,
            title: track.name,
            uri: track.uri,
            albumUrl: smallesAlbumImage.url,
          }
        })
      )
    })

    return () => (cancel = true)
  }, [search, accessToken])

  const handleSearch = e => {
    setSearch(e.target.value)
  }

  return (
    <Container>
      <Grid2>
        <div>
          <Form>
            <SearchBar
              type='search'
              placeholder='Search Songs'
              value={search}
              onChange={handleSearch}
            />
          </Form>

          <TrackSearchResultContainer>
            {searchResults.map(track => {
              return (
                <SearchResults
                  track={track}
                  key={track.uri}
                  chooseTrack={chooseTrack}
                />
              )
            })}
          </TrackSearchResultContainer>
        </div>

        <div>
          <Player accessToken={accessToken} trackUri={playingTrack?.uri} />
        </div>
      </Grid2>
    </Container>
    
  )
}

export default Dashboard
