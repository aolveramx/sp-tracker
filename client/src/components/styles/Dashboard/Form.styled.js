import styled from 'styled-components'

export const Form = styled.form`
  align-items: center;
  background-color: #37474f;
  border-radius: 10rem;
  cursor: pointer;
  display: flex;
  height: 2rem;
  justify-content: center;
  margin-top: 1rem;
  padding: 1rem;
  position: relative;
  width: 30rem;
`