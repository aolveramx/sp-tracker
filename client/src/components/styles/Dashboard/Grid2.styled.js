import styled from 'styled-components'

export const Grid2 = styled.div`
  display: grid;
  grid-template-columns: repeat(2, 1fr);
  grid-gap: 2rem;
  height: 100vh;

  @media(max-width: 1045px) {
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
  }
`