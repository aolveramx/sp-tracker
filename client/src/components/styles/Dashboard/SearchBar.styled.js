import styled from 'styled-components'

export const SearchBar = styled.input`
  background-color: transparent;
  border: none;
  color: #fff;
  font-size: 14px;
  line-height: 1;
  padding-left: 1rem;
  width: 100%;

  &:focus,
  &:active {
    outline: none;
  }
  &::placeholder {
    color: white;
  }
`