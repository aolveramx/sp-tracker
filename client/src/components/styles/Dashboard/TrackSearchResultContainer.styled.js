import styled from 'styled-components'

export const TrackSearchResultContainer = styled.div`
  flex-grow: 1;
  margin-top: 2rem;
  overflow-y: auto;
`