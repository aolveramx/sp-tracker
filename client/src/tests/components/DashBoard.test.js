import '@testing-library/jest-dom'
import { shallow } from 'enzyme'
import Dashboard from '../../components/Dashboard';


describe('Tests in Dashboard.js', () => {

  test('should render <Dashboard />', () => {

    const wrapper = shallow(<Dashboard />)
    expect(wrapper).toMatchSnapshot()
    
  });
  
});