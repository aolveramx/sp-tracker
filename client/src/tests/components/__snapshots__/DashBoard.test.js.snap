// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in Dashboard.js should render <Dashboard /> 1`] = `
<styled.div>
  <styled.div>
    <div>
      <styled.form>
        <styled.input
          onChange={[Function]}
          placeholder="Search Songs"
          type="search"
          value=""
        />
      </styled.form>
      <styled.div />
    </div>
    <div>
      <Player />
    </div>
  </styled.div>
</styled.div>
`;
